ipa-core/src/protocol/ipa_prf/shuffle/malicious.rs

```diff
@@ -0,0 +1,237 @@
+use futures::stream::StreamExt;
+use futures_util::stream::iter;
+use generic_array::GenericArray;
+
+use crate::{
+    error::Error,
+    ff::{boolean_array::BooleanArray, Gf32Bit, Serializable},
+    helpers::TotalRecords,
+    protocol::{basics::mul::semi_honest_multiply, context::Context, RecordId},
+    secret_sharing::replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+    seq_join::seq_join,
+};
+
+/// This function computes the MAC tag for each row and appends it to the row.
+/// It outputs the vector of rows concatenated with the tags.
+///
+/// The tag is the inner product between keys and row entries,
+/// i.e. `Sum_i key_i * row_entry_i`.
+///
+/// The multiplication is in `Gf32Bit`.
+/// Therefore, each row is split into `32 bit` row entries
+///
+/// ## Error
+/// Propagates MPC multiplication errors.
+///
+/// ## Panics
+/// When conversion fails or when `S::Bits + 32 != B::Bits`.
+async fn compute_and_add_tags<C: Context, S: BooleanArray, B: BooleanArray>(
+    ctx: C,
+    keys: &[AdditiveShare<Gf32Bit>],
+    rows: &[AdditiveShare<S>],
+) -> Result<Vec<AdditiveShare<B>>, Error> {
+    let length = rows.len();
+    let row_length = keys.len();
+    let tag_ctx = ctx.set_total_records(TotalRecords::specified(length * row_length)?);
```

Comment:
this will blow up if arrays are of size 0. It is fine to check these outside of this function, just make sure the check is there somewhere - we had already a case where shuffle received 0 rows and failed in a random place.

I would generally just slap `debug_assert` if I expect that check to be somewhere up the stack